File: case2/c2_main.py
#!/usr/bin/env python3
# Technique: Case 2 - Minimum Jerk QP Trajectory Planning + 6-Parameter Pseudo-Flatness Reconstruction (m11 != m22)

import os
import time
import json
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

from c2_usv_params import DT_SIM, T_MAX, T_MIN
from c2_min_jerk_qp import MinJerkTrajectory2D
from c2_flatness_reconstruct import reconstruct_flatness_h2

plt.rcParams['font.sans-serif'] = 'DejaVu Sans'
plt.rcParams['axes.edgecolor'] = '#333333'
plt.rcParams['axes.linewidth'] = 1.2

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    waypoints = np.array([
        [0.0, 0.0],
        [8.0, 0.0],
        [14.0, 5.0],
        [14.0, 13.0],
        [20.0, 17.0],
        [28.0, 17.0],
        [32.0, 10.0],
        [26.0, 4.0],
        [20.0, 1.5]
    ])
    
    base_times = np.array([0.0, 7.0, 14.0, 20.0, 27.0, 33.0, 40.0, 48.0, 55.0])
    time_scale = 2.10
    times = base_times * time_scale
    
    v0_vec = (0.1, 0.0)
    dir_f = waypoints[8] - waypoints[7]
    dir_f_unit = dir_f / np.linalg.norm(dir_f)
    vf_vec = tuple(0.01 * dir_f_unit)
    
    t_start = time.perf_counter()
    planner = MinJerkTrajectory2D(waypoints, times, vel_start=v0_vec, vel_end=vf_vec)
    t_sim, pos, vel, acc, jerk = planner.sample(dt=DT_SIM)
    flat_data = reconstruct_flatness_h2(pos, vel, acc, jerk, t_sim)
    solve_time_ms = (time.perf_counter() - t_start) * 1000.0

    
    eta_ref  = flat_data['eta']
    nu_ref   = flat_data['nu']
    tau_ref  = flat_data['tau_plan']
    T_plan   = flat_data['T_plan']
    
    planning_metrics = {
        "case": "Case 2",
        "solver_type": "QP (6-Param Pseudo-Flatness)",
        "solve_time_ms": float(solve_time_ms),
        "total_sim_time_s": float(t_sim[-1] - t_sim[0]),
        "num_samples": int(len(t_sim))
    }
    metrics_file = os.path.join(script_dir, 'planning_metrics.json')
    with open(metrics_file, 'w') as f:
        json.dump(planning_metrics, f, indent=4)
    print(f"[Main] Trajectory Solver (QP) Compute Time: {solve_time_ms:.4f} ms")

    fig, axs = plt.subplots(5, 1, figsize=(10, 16), dpi=300)
    c_blue  = '#2563EB'
    c_green = '#059669'
    c_amber = '#D97706'
    c_red   = '#DC2626'
    
    axs[0].plot(eta_ref[:, 0], eta_ref[:, 1], color=c_blue, lw=2.2, label='Planned Path')
    axs[0].scatter(waypoints[:, 0], waypoints[:, 1], color='#111827', s=55, zorder=5, label='Waypoints')
    for idx_wp, (wx, wy) in enumerate(waypoints):
        axs[0].annotate(f'WP{idx_wp}', (wx, wy), textcoords="offset points", xytext=(5, 5), fontsize=8, fontweight='bold')
    axs[0].set_xlabel('X [m]', fontweight='bold')
    axs[0].set_ylabel('Y [m]', fontweight='bold')
    axs[0].set_title('2D Trajectory', fontsize=12, fontweight='bold', pad=8)
    axs[0].grid(True, ls=':', alpha=0.6)
    axs[0].axis('equal')
    axs[0].legend(loc='best', frameon=True, facecolor='white')
    
    axs[1].plot(t_sim, nu_ref[:, 0], color=c_blue, lw=1.8, label='$u$ [m/s]')
    axs[1].plot(t_sim, nu_ref[:, 1], color=c_green, lw=1.8, label='$v$ [m/s]')
    axs[1].plot(t_sim, nu_ref[:, 2], color=c_amber, lw=1.8, label='$r$ [rad/s]')
    axs[1].set_xlabel('Time [s]', fontweight='bold')
    axs[1].set_ylabel('Velocities', fontweight='bold')
    axs[1].set_title('Body Velocities', fontsize=12, fontweight='bold', pad=8)
    axs[1].grid(True, ls=':', alpha=0.6)
    axs[1].legend(loc='upper right', ncol=3, frameon=True, facecolor='white', fontsize=9)
    
    axs[2].plot(t_sim, tau_ref[:, 0], color=c_blue, lw=1.8, label='$\\tau_u$ [N]')
    axs[2].plot(t_sim, tau_ref[:, 1], color=c_amber, lw=1.8, label='$\\tau_r$ [N·m]')
    axs[2].set_xlabel('Time [s]', fontweight='bold')
    axs[2].set_ylabel('Forces', fontweight='bold')
    axs[2].set_title('Control Forces', fontsize=12, fontweight='bold', pad=8)
    axs[2].grid(True, ls=':', alpha=0.6)
    axs[2].legend(loc='upper right', ncol=2, frameon=True, facecolor='white', fontsize=9)
    
    axs[3].plot(t_sim, T_plan[:, 0], color=c_green, lw=1.8, label='$T_1$ [N]')
    axs[3].plot(t_sim, T_plan[:, 1], color=c_red, lw=1.8, label='$T_2$ [N]')
    axs[3].axhline(T_MAX, color='#9CA3AF', ls=':', lw=1.2, label=f'$T_{{max}}$ ({T_MAX:.1f} N)')
    axs[3].axhline(T_MIN, color='#9CA3AF', ls=':', lw=1.2, label=f'$T_{{min}}$ ({T_MIN:.1f} N)')
    axs[3].set_xlabel('Time [s]', fontweight='bold')
    axs[3].set_ylabel('Thrust [N]', fontweight='bold')
    axs[3].set_title('Thruster Allocation', fontsize=12, fontweight='bold', pad=8)
    axs[3].grid(True, ls=':', alpha=0.6)
    axs[3].legend(loc='upper right', ncol=4, frameon=True, facecolor='white', fontsize=8.5)
    
    jerk_mag = np.hypot(jerk[:, 0], jerk[:, 1])
    axs[4].plot(t_sim, jerk[:, 0], color=c_blue, lw=1.5, label='$j_x$ [m/s³]')
    axs[4].plot(t_sim, jerk[:, 1], color=c_green, lw=1.5, label='$j_y$ [m/s³]')
    axs[4].plot(t_sim, jerk_mag, color=c_red, lw=1.8, ls='-', label='$|j|$ [m/s³]')
    axs[4].set_xlabel('Time [s]', fontweight='bold')
    axs[4].set_ylabel('Jerk [m/s³]', fontweight='bold')
    axs[4].set_title('Jerk Profiles', fontsize=12, fontweight='bold', pad=8)
    axs[4].grid(True, ls=':', alpha=0.6)
    axs[4].legend(loc='upper right', ncol=3, frameon=True, facecolor='white', fontsize=9)
    
    plt.tight_layout()
    out_img_path = os.path.join(script_dir, 'c2_flatness_planning_results.png')
    plt.savefig(out_img_path, dpi=300, bbox_inches='tight')
    plt.close()
    print(f"[Main] Planning completed. Diagnostic plot saved to: {out_img_path}")

if __name__ == '__main__':
    main()

File: case2_semilla_c/src/c2_main.c
/* Technique: Case 2 - Minimum Jerk QP Trajectory Planning +
 * 6-Parameter Pseudo-Flatness Reconstruction (m11 != m22)
 * C port of c2_main.py (planning only, no closed/open-loop simulation). */

#define _POSIX_C_SOURCE 199309L
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#include "c2_usv_params.h"
#include "c2_min_jerk_qp.h"
#include "c2_flatness_reconstruct.h"

#define N_WP 9

static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}

int main(void) {
    double waypoints_x[N_WP] = {0.0, 8.0, 14.0, 14.0, 20.0, 28.0, 32.0, 26.0, 20.0};
    double waypoints_y[N_WP] = {0.0, 0.0, 5.0, 13.0, 17.0, 17.0, 10.0, 4.0, 1.5};
    double base_times[N_WP]  = {0.0, 7.0, 14.0, 20.0, 27.0, 33.0, 40.0, 48.0, 55.0};
    double time_scale = 2.10;
    double times[N_WP];
    for (int i = 0; i < N_WP; i++) times[i] = base_times[i] * time_scale;

    double v0x = 0.1, v0y = 0.0;
    double dfx = waypoints_x[N_WP - 1] - waypoints_x[N_WP - 2];
    double dfy = waypoints_y[N_WP - 1] - waypoints_y[N_WP - 2];
    double dn = hypot(dfx, dfy);
    double vfx = 0.01 * dfx / dn, vfy = 0.01 * dfy / dn;

    double t_start_ms = now_ms();

    MinJerk2D planner;
    minjerk2d_solve(&planner, waypoints_x, waypoints_y, times, N_WP, v0x, v0y, vfx, vfy);

    double *t_sim, *px, *py, *vx, *vy, *ax, *ay, *jx, *jy;
    int N;
    minjerk2d_sample(&planner, DT_SIM, &t_sim, &px, &py, &vx, &vy, &ax, &ay, &jx, &jy, &N);

    FlatnessResult flat;
    reconstruct_flatness_h2(N, t_sim, px, py, vx, vy, ax, ay, &flat);

    double solve_time_ms = now_ms() - t_start_ms;

    printf("[Main] Trajectory Solver (QP) Compute Time: %.4f ms\n", solve_time_ms);

    /* --- c2_waypoints.csv --- */
    FILE *fwp = fopen("c2_waypoints.csv", "w");
    if (fwp) {
        fprintf(fwp, "idx,x,y,t\n");
        for (int i = 0; i < N_WP; i++)
            fprintf(fwp, "%d,%.6f,%.6f,%.6f\n", i, waypoints_x[i], waypoints_y[i], times[i]);
        fclose(fwp);
    }

    /* --- c2_main_results.csv --- */
    FILE *fr = fopen("c2_main_results.csv", "w");
    if (fr) {
        fprintf(fr, "t,x,y,psi,u,v,r,tau_u,tau_r,T1,T2,cmd1,cmd2,jerk_x,jerk_y\n");
        for (int i = 0; i < N; i++)
            fprintf(fr, "%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n",
                    t_sim[i], flat.x[i], flat.y[i], flat.psi[i],
                    flat.u[i], flat.v[i], flat.r[i],
                    flat.tau_u[i], flat.tau_r[i],
                    flat.T1_dem[i], flat.T2_dem[i],
                    flat.cmd1[i], flat.cmd2[i],
                    jx[i], jy[i]);
        fclose(fr);
    }

    /* --- planning_metrics.json --- */
    FILE *fm = fopen("planning_metrics.json", "w");
    if (fm) {
        fprintf(fm,
            "{\n"
            "    \"case\": \"Case 2\",\n"
            "    \"solver_type\": \"QP (6-Param Pseudo-Flatness) - C port\",\n"
            "    \"solve_time_ms\": %.6f,\n"
            "    \"total_sim_time_s\": %.6f,\n"
            "    \"num_samples\": %d\n"
            "}\n",
            solve_time_ms, t_sim[N - 1] - t_sim[0], N);
        fclose(fm);
    }

    printf("[Main] Planning completed.\n");
    printf("  c2_waypoints.csv, c2_main_results.csv, planning_metrics.json written.\n");
    printf("  Run 'python3 plot_results.py' to regenerate the diagnostic PNGs.\n");

    flatness_result_free(&flat);
    free(t_sim); free(px); free(py); free(vx); free(vy); free(ax); free(ay); free(jx); free(jy);
    minjerk2d_free(&planner);
    return 0;
}
